RecoParticleFlow/PFClusterProducer/interface/PFRecHitQTests.h

```diff
@@ -292,16 +305,26 @@ class PFRecHitQTestHCALThresholdVsDepth : public PFRecHitQTestBase {
   std::vector<std::vector<int>> depths_;
   std::vector<std::vector<double>> thresholds_;
   std::vector<int> detector_;
+  const HcalTopology* htopo;
+  const HcalPFCuts* hcalCuts;
 
   bool test(unsigned aDETID, double energy, double time, bool& clean) {
     HcalDetId detid(aDETID);
+    const HcalPFCut* item = nullptr;
+    if (cutsFromDB) {
+      std::unique_ptr<HcalPFCuts> paramPF_;
+      paramPF_ = std::make_unique<HcalPFCuts>(*hcalCuts);
+      paramPF_->setTopo(htopo);
```

Comment:
This pattern of creating a unique pointer to `HcalPFCuts` and then consuming the topology over end over, here and in other PRs. Should this be created only once in a `ESProducer` and be consumed as event setup product? @cms-sw/core-l2 please advice

Reply:
I agree, it would make sense to move this topology-set-`HcalPFCuts` into EventSetup. The new `HcalPFCuts` data product would have to be put in a new(?) Record that depends on `HcalPFCutsRcd` and `HcalRecNumberingRecord`.

The possibility of using `HcalPFCuts` without the topology (even if the code has checks) is not very good. A better alternative for the current class structure would be to leave the `COND_SERIALIZABLE` data members as they are, and craft a new EventSetup data product class provides the interface for users (that is currently in `HcalCondObjectContainer`) while keeping pointers to the data-containing objects, such that user would be able to interact only with properly initialized (with the topology) data product.

Reply:
for this, I guess @cms-sw/alca-l2 and @cms-sw/hcal-dpg-l2 would need to step in and help out. 
also tagging @abdoulline and @igv4321

Reply:
Before embarking upon changing HcalCondObjectContainer and and struggling with the real possibility of introducing new bugs, it would be nice to hear a realistic scenario in which HcalCondObjectContainer can fail. The current design (in which HcalCondObjectContainer can work either with or without the topology pointer) caused no problems whatsoever for the past ten years or so. This is precisely why the checks for the pointer are included (technically, inside HcalCondObjectContainerBase) in all methods that depend on the presence of this pointer.

Also, how many modules we are talking about that call "setTopo"? This is done once per run, so these calls are expected to consume negligible amount of resources.

I agree that employing std::unique_ptr in this manner looks pointless, as the code does not attempt to handle any exceptions. Instead, it would make more sense to make paramPF_ a class member, and then one does not have to hold the object by the bare pointer and deallocate manually inside endRun.

Reply:
> Also, how many modules we are talking about that call "setTopo" after copying HcalPFCuts? This is done once per run, so these calls are expected to consume negligible amount of resources.

summing up this PR (https://github.com/cms-sw/cmssw/pull/43025) + https://github.com/cms-sw/cmssw/pull/43176 + https://github.com/cms-sw/cmssw/pull/43164 I count a total of 6 different instances.
I agree the cost should be small (provided the threshold DB object doesn't become massive in future for whatever reason). On the other hand seems a bit inelegant. Is it possible to follow the original proposal to wrap the topology-set `HCalPFCuts` in a single ESproduct without changing `HcalCondObjectContainer` ? I let Matti to elaborate on the longer term proposal to alter the `HcalCondObjectContainer` design.

Reply:
> I let Matti to elaborate on the longer term proposal to alter the `HcalCondObjectContainer` design.

Currently the mistake of calling functions that need the topology without the call to `setTopo()` can be spotted only at run time. That change would prevent those kind of mistakes already at compile time. I don't really know what to elaborate further from https://github.com/cms-sw/cmssw/pull/43025#discussion_r1382214112 without specific questions.

Reply:
Are you serious? If yes, let's scrap all that python infrastructure we have around CMSSW and switch to pure compiled code. That way, after 100 years or so of development effort, we can build a system in which all errors could be caught by the compiler.

Reply:
It was merely a suggestion, not a required change (at leas by me). Feel free to ignore.